File: src/OpenSilver.Showcase/Other/Internals/ItemsControlLoadingBehavior.vb
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Controls.Primitives

Namespace OpenSilver.Showcase
    Public NotInheritable Class ItemsControlLoadingBehavior
        Private Sub New()
        End Sub

        Public Shared ReadOnly ShowLoadingOnGeneratingProperty As DependencyProperty =
            DependencyProperty.RegisterAttached(
                "ShowLoadingOnGenerating",
                GetType(Boolean),
                GetType(ItemsControlLoadingBehavior),
                New PropertyMetadata(False, AddressOf OnShowLoadingOnGeneratingChanged)
            )

        Public Shared Sub SetShowLoadingOnGenerating(element As DependencyObject, value As Boolean)
            element.SetValue(ShowLoadingOnGeneratingProperty, value)
        End Sub

        Public Shared Function GetShowLoadingOnGenerating(element As DependencyObject) As Boolean
            Return CBool(element.GetValue(ShowLoadingOnGeneratingProperty))
        End Function

        Private Shared ReadOnly StatusChangedHandlerProperty As DependencyProperty =
            DependencyProperty.RegisterAttached(
                "StatusChangedHandler",
                GetType(EventHandler),
                GetType(ItemsControlLoadingBehavior),
                Nothing
            )

        Private Shared ReadOnly LoadingPopupProperty As DependencyProperty =
            DependencyProperty.RegisterAttached(
                "LoadingPopup",
                GetType(Popup),
                GetType(ItemsControlLoadingBehavior),
                Nothing
            )

        Private Shared Sub OnShowLoadingOnGeneratingChanged(d As DependencyObject, e As DependencyPropertyChangedEventArgs)
            Dim itemsControl = TryCast(d, ItemsControl)
            If itemsControl Is Nothing Then Return

            Dim enabled As Boolean = CBool(e.NewValue)
            If enabled Then
                Dim handler As EventHandler = Sub(s, args) OnStatusChanged(itemsControl)
                itemsControl.SetValue(StatusChangedHandlerProperty, handler)
                AddHandler itemsControl.ItemContainerGenerator.StatusChanged, handler
            Else
                Dim handler = TryCast(itemsControl.GetValue(StatusChangedHandlerProperty), EventHandler)
                If handler IsNot Nothing Then
                    RemoveHandler itemsControl.ItemContainerGenerator.StatusChanged, handler
                    itemsControl.ClearValue(StatusChangedHandlerProperty)
                End If

                Dim existing = TryCast(itemsControl.GetValue(LoadingPopupProperty), Popup)
                If existing IsNot Nothing Then
                    existing.IsOpen = False
                    itemsControl.ClearValue(LoadingPopupProperty)
                End If
            End If
        End Sub

        Private Shared Sub OnStatusChanged(itemsControl As ItemsControl)
            Dim status = itemsControl.ItemContainerGenerator.Status

            If status = GeneratorStatus.GeneratingContainers Then
                Dim popup As New Popup With {
                    .Child = New LoadingControl(),
                    .Placement = PlacementMode.Absolute,
                    .IsHitTestVisible = False,
                    .IsOpen = False
                }

                ' Overlay the whole window
                Dim host = TryCast(Application.Current.RootVisual, FrameworkElement)
                popup.Width = host.ActualWidth
                popup.Height = host.ActualHeight

                itemsControl.SetValue(LoadingPopupProperty, popup)
                popup.IsOpen = True
            Else
                Dim popup = TryCast(itemsControl.GetValue(LoadingPopupProperty), Popup)
                If popup IsNot Nothing Then
                    popup.IsOpen = False
                    itemsControl.ClearValue(LoadingPopupProperty)
                End If
            End If
        End Sub
    End Class
End Namespace

File: src/OpenSilver.Showcase/Other/Internals/ItemsControlLoadingBehavior.fs
namespace OpenSilver.Showcase

open System
open System.Windows
open System.Windows.Controls
open System.Windows.Controls.Primitives

type ItemsControlLoadingBehavior() =

    static let statusChangedHandlerProperty =
        DependencyProperty.RegisterAttached(
            "StatusChangedHandler",
            typeof<bool>,
            typeof<ItemsControlLoadingBehavior>,
            null
        )

    static let loadingPopupProperty =
        DependencyProperty.RegisterAttached(
            "LoadingPopup",
            typeof<Popup>,
            typeof<ItemsControlLoadingBehavior>,
            null
        )

    static let rec onStatusChanged (itemsControl: ItemsControl) =
        match itemsControl.ItemContainerGenerator.Status with
        | GeneratorStatus.GeneratingContainers ->
            let popup = Popup()
            popup.Child <- new LoadingControl()
            popup.Placement <- PlacementMode.Absolute
            popup.IsHitTestVisible <- false
            popup.IsOpen <- false

            let host = Application.Current.RootVisual :?> FrameworkElement
            popup.Width <- host.ActualWidth
            popup.Height <- host.ActualHeight

            itemsControl.SetValue(loadingPopupProperty, popup)
            popup.IsOpen <- true
        | _ ->
            match itemsControl.GetValue(loadingPopupProperty) with
            | :? Popup as popup ->
                popup.IsOpen <- false
                itemsControl.ClearValue(loadingPopupProperty)
            | _ -> ()

    static let ShowLoadingOnGeneratingProperty =
        DependencyProperty.RegisterAttached(
            "ShowLoadingOnGenerating",
            typeof<bool>,
            typeof<ItemsControlLoadingBehavior>,
            PropertyMetadata(false, PropertyChangedCallback(fun d e ->
                match d with
                | :? ItemsControl as ic ->
                    let enabled = e.NewValue :?> bool
                    if enabled then
                        let handler = EventHandler(fun _ _ -> onStatusChanged ic)
                        ic.SetValue(statusChangedHandlerProperty, handler)
                        ic.ItemContainerGenerator.StatusChanged.AddHandler(handler)
                    else
                        match ic.GetValue(statusChangedHandlerProperty) with
                        | :? EventHandler as handler ->
                            ic.ItemContainerGenerator.StatusChanged.RemoveHandler(handler)
                            ic.ClearValue(statusChangedHandlerProperty)
                        | _ -> ()

                        match ic.GetValue(loadingPopupProperty) with
                        | :? Popup as popup ->
                            popup.IsOpen <- false
                            ic.ClearValue(loadingPopupProperty)
                        | _ -> ()
                | _ -> ()
            ))
        )

    static member SetShowLoadingOnGenerating (element: DependencyObject, value: bool) =
        element.SetValue(ShowLoadingOnGeneratingProperty, value)

    static member GetShowLoadingOnGenerating (element: DependencyObject) =
        element.GetValue(ShowLoadingOnGeneratingProperty) :?> bool
